gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -0,0 +1,897 @@
+        VSIFReadL(m_featureOffsets, 8, featuresCount, m_poFp);
+    } else {
+        CPLDebug("FlatGeobuf", "processSpatialIndex noop");
+    }
+}
+
+GIntBig OGRFlatGeobufLayer::GetFeatureCount(int bForce) {
+    if (m_poFilterGeom != nullptr || m_poAttrQuery != nullptr) {
+        return OGRLayer::GetFeatureCount(bForce);
+    } else {
+        return m_featuresCount;
+    }
+}
+
+OGRFeature *OGRFlatGeobufLayer::GetNextFeature()
+{
+    while( true ) {
+        if (m_featuresPos >= m_featuresCount) {
+            CPLDebug("FlatGeobuf", "GetNextFeature: iteration end");
+            if (m_poFp != nullptr) {
+                VSIFCloseL(m_poFp);
+                m_poFp = nullptr;
+            }
+            return nullptr;
+        }
+
+        if (m_poFp == nullptr) {
+            CPLDebug("FlatGeobuf", "GetNextFeature: iteration start (will attempt to open file %s)", m_pszFilename);
+            m_poFp = VSIFOpenL(m_pszFilename, "rb");
+            if (m_poFp == nullptr) {
+                CPLError(CE_Failure, CPLE_AppDefined, "Failed to open file");
+                return nullptr;
+            }
+            //m_poFp = (VSILFILE*) VSICreateCachedFile ( (VSIVirtualHandle*) m_poFp);
+
+            processSpatialIndex();
+            if (m_featuresCount == 0) {
+                CPLDebug("FlatGeobuf", "GetNextFeature: no features found in spatial index search");
+                if (m_poFp != nullptr) {
+                    VSIFCloseL(m_poFp);
+                    m_poFp = nullptr;
+                }
+                return nullptr;
+            }
+        }
+
+        if (m_processedSpatialIndex && !m_ignoreSpatialFilter)
+            m_offset = m_offsetInit + m_featureOffsets[m_foundFeatureIndices[m_featuresPos]];
+        else if (m_featuresPos > 0)
+            m_offset += m_featureSize + sizeof(uoffset_t);
+
+        OGRFeature *poFeature = new OGRFeature(m_poFeatureDefn);
+
+        VSIFSeekL(m_poFp, m_offset, SEEK_SET);
+        VSIFReadL(&m_featureSize, sizeof(uoffset_t), 1, m_poFp);
+        CPL_LSBPTR32(&m_featureSize);
+        if (m_featureBufSize == 0) {
+            m_featureBufSize = std::max(1024U * 32U, m_featureSize);
+            CPLDebug("FlatGeobuf", "GetNextFeature: m_featureBufSize: %d", m_featureBufSize);
+            m_featureBuf = static_cast<GByte *>(VSIMalloc(m_featureBufSize));
+        } else if (m_featureBufSize < m_featureSize) {
+            m_featureBufSize = std::max(m_featureBufSize * 2, m_featureSize);
+            CPLDebug("FlatGeobuf", "GetNextFeature: m_featureBufSize: %d", m_featureBufSize);
+            m_featureBuf = static_cast<GByte *>(VSIRealloc(m_featureBuf, m_featureBufSize));
+        }
+        VSIFReadL(m_featureBuf, 1, m_featureSize, m_poFp);
+
+    #ifdef DEBUG
+        const uint8_t * vBuf = const_cast<const uint8_t *>(reinterpret_cast<uint8_t *>(m_featureBuf));
+        Verifier v(vBuf, m_featureSize);
+        auto ok = VerifyFeatureBuffer(v);
```

Comment:
what would happen in non-DEBUG mode when VerifyFeatureBuffer() isn't run and if the buffer is corrupted ?

Reply:
Undefined behaviour. It's fairly expensive to verify the buffer so I wouldn't want that to be mandatory in a performance oriented format.

Reply:
Well, we try hard to secure GDAL against undefined (hostile) behaviour in its default configuration. Are you sure those tests are really costly ? Couldn't the checks be let by default and skipped if the user defines some configuration option at runtime and/or a #define at build time ?

Reply:
A runtime opt out is reasonable, I'll add that. Just have to be a bit careful to not promise 100% safety regardless when documenting the behavior.

Reply:
Introduced runtime opt out (VERIFY_BUFFERS option).